presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java

```diff
@@ -346,4 +354,95 @@ public SettableFuture schedule(SourcePartitionedScheduler scheduler)
             return newDriverGroupReady;
         }
     }
+
+    private abstract static class SubScheduler
+    {
+        final SourcePartitionedScheduler sourcePartitionedScheduler;
+
+        public SubScheduler(SourcePartitionedScheduler sourcePartitionedScheduler)
+        {
+            this.sourcePartitionedScheduler = sourcePartitionedScheduler;
+        }
+
+        public ScheduleResult schedule()
+        {
+            return sourcePartitionedScheduler.schedule();
+        }
+
+        public void close()
+        {
+            sourcePartitionedScheduler.close();
+        }
+
+        public PlanNodeId getPlanNodeId()
+        {
+            return sourcePartitionedScheduler.getPlanNodeId();
+        }
+
+        public abstract void startLifespan(Lifespan lifespan, ConnectorPartitionHandle partitionHandle);
+
+        public abstract List<Lifespan> drainCompletedLifespans();
+    }
+
+    private static class SimpleSubScheduler
+            extends SubScheduler
+    {
+        public SimpleSubScheduler(SourcePartitionedScheduler sourcePartitionedScheduler)
+        {
+            super(sourcePartitionedScheduler);
+        }
+
+        @Override
+        public void startLifespan(Lifespan lifespan, ConnectorPartitionHandle partitionHandle)
+        {
+            sourcePartitionedScheduler.startLifespan(lifespan, partitionHandle);
+        }
+
+        @Override
+        public List<Lifespan> drainCompletedLifespans()
+        {
+            return sourcePartitionedScheduler.drainCompletedLifespans();
+        }
+    }
+
+    private static class AdaptedSubScheduler
```

Comment:
What's an "adapted" sub scheduler? We need a name that better conveys its purpose.